File: examples/typescript/discord/discord_fuse.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { readdir, readFile } from "node:fs/promises";
import { createInterface } from "node:readline/promises";
import dotenv from "dotenv";
import {
  DiscordResource,
  Mount,
  MountMode,
  Workspace,
  type DiscordConfig,
} from "@struktoai/mirage-node";

dotenv.config({ path: ".env.development" });

function buildConfig(): DiscordConfig {
  const token = process.env.DISCORD_BOT_TOKEN;
  if (token === undefined || token === "") {
    throw new Error("DISCORD_BOT_TOKEN env var is required");
  }
  return { token };
}

async function main(): Promise<void> {
  const resource = new DiscordResource(buildConfig());
  const ws = new Workspace({
    "/discord": new Mount(resource, { mode: MountMode.READ, fuse: true }),
  });
  await ws.fuseReady();
  const mp = ws.fuseMountpoint as string;
  try {
    console.log(`=== FUSE MODE: mounted at ${mp} ===\n`);

    console.log("--- readdir() guilds ---");
    const guilds = await readdir(mp);
    for (const g of guilds) {
      console.log(`  ${g}`);
    }

    if (guilds.length === 0) {
      console.log("no guilds found");
    } else {
      const guild = guilds[0]!;

      console.log(`\n--- readdir() ${guild} ---`);
      const contents = await readdir(`${mp}/${guild}`);
      for (const c of contents) {
        console.log(`  ${c}`);
      }

      console.log(`\n--- readdir() ${guild}/channels ---`);
      const channels = await readdir(`${mp}/${guild}/channels`);
      for (const ch of channels) {
        console.log(`  ${ch}`);
      }

      if (channels.length > 0) {
        const ch = channels[0]!;

        console.log(`\n--- readdir() ${ch} (last 5 dates) ---`);
        const dates = await readdir(`${mp}/${guild}/channels/${ch}`);
        for (const d of dates.slice(-5)) {
          console.log(`  ${d}`);
        }

        if (dates.length > 0) {
          const target = dates[dates.length - 1]!;
          const dateDir = `${mp}/${guild}/channels/${ch}/${target}`;
          const chatPath = `${dateDir}/chat.jsonl`;
          console.log(`\n--- readFile() ${target}/chat.jsonl ---`);
          const text = (await readFile(chatPath, "utf-8")).trim();
          if (text !== "") {
            const lines = text.split("\n").filter((ln) => ln.trim() !== "");
            for (let i = 0; i < lines.length; i++) {
              if (i >= 5) {
                console.log("  ...");
                break;
              }
              try {
                const msg = JSON.parse(lines[i]!) as {
                  author?: { username?: string };
                  content?: string;
                };
                const author = msg.author?.username ?? "?";
                const content = (msg.content ?? "").slice(0, 80);
                console.log(`  [${author}] ${content}`);
              } catch {
                break;
              }
            }
          } else {
            console.log("  (empty — no messages on this date)");
          }
          try {
            const atts = await readdir(`${dateDir}/files`);
            if (atts.length > 0) {
              console.log(`\n--- readdir() ${target}/files ---`);
              for (const a of atts.slice(0, 5)) console.log(`  ${a}`);
            }
          } catch {
            // no files dir
          }
        }
      }

      console.log(`\n--- readdir() ${guild}/members ---`);
      const members = await readdir(`${mp}/${guild}/members`);
      for (const m of members.slice(0, 5)) {
        console.log(`  ${m}`);
      }

      if (members.length > 0) {
        const memberPath = `${mp}/${guild}/members/${members[0]!}`;
        console.log(`\n--- readFile() ${members[0]!} ---`);
        const text = (await readFile(memberPath, "utf-8")).trim();
        if (text !== "") {
          try {
            const data = JSON.parse(text) as {
              user?: { username?: string; id?: string };
            };
            console.log(`  username: ${String(data.user?.username)}`);
            console.log(`  id: ${String(data.user?.id)}`);
          } catch {
            console.log(`  (raw: ${text.slice(0, 100)})`);
          }
        } else {
          console.log("  (empty)");
        }
      }
    }

    console.log(`\n>>> FUSE mounted at: ${mp}`);
    console.log(">>> Open another terminal and run:");
    console.log(`>>>   ls ${mp}/`);
    console.log(`>>>   cat ${mp}/<guild>/channels/<ch>/<date>/chat.jsonl`);
    console.log(">>> Press Enter to unmount and exit...");

    const rl = createInterface({
      input: process.stdin,
      output: process.stdout,
    });
    await rl.question("");
    rl.close();

    const records = ws.records;
    const total = records.reduce((acc, r) => acc + (r.bytes ?? 0), 0);
    console.log(
      `\nStats: ${String(records.length)} ops, ${String(total)} bytes transferred`,
    );
  } finally {
    await ws.close();
  }
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});

File: examples/python/discord/discord_fuse.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
import os
import subprocess

from dotenv import load_dotenv

from mirage import Mount, MountMode, Workspace
from mirage.resource.discord import DiscordConfig, DiscordResource

load_dotenv(".env.development")

config = DiscordConfig(token=os.environ["DISCORD_BOT_TOKEN"])
resource = DiscordResource(config=config)

with Workspace({"/discord/": Mount(resource, mode=MountMode.READ,
                                   fuse=True)}) as ws:
    mp = ws.fuse_mountpoint

    print(f"=== FUSE MODE: mounted at {mp} ===\n")

    # ── list guilds ──────────────────────────────
    print("--- os.listdir() guilds ---")
    # Skip the virtual /.mirage dir (agent metadata), it is not a guild.
    guilds = [g for g in os.listdir(mp) if not g.startswith(".")]
    for g in guilds:
        print(f"  {g}")

    if not guilds:
        print("no guilds found")
    else:
        guild = guilds[0]

        # ── list guild contents ──────────────────
        print(f"\n--- os.listdir() {guild} ---")
        contents = os.listdir(f"{mp}/{guild}")
        for c in contents:
            print(f"  {c}")

        # ── list channels ────────────────────────
        print(f"\n--- os.listdir() {guild}/channels ---")
        channels = os.listdir(f"{mp}/{guild}/channels")
        for ch in channels:
            print(f"  {ch}")

        if channels:
            ch = channels[0]

            # ── list date files ──────────────────
            print(f"\n--- os.listdir() {ch} (last 5 dates) ---")
            dates = os.listdir(f"{mp}/{guild}/channels/{ch}")
            for d in dates[-5:]:
                print(f"  {d}")

            # ── read chat.jsonl ──────────────────
            if dates:
                target = dates[-1]
                date_dir = f"{mp}/{guild}/channels/{ch}/{target}"
                chat_path = f"{date_dir}/chat.jsonl"
                # chat.jsonl has no size until fetched: unopened it stats as
                # 0 bytes; any open (cat/wc/cp) hydrates it, and stat then
                # reports the real size (see docs/python/setup/fuse.mdx).
                print(
                    f"\n--- size-unknown semantics on {target}/chat.jsonl ---")
                print(
                    f"  stat before open: {os.stat(chat_path).st_size} bytes")
                wc = subprocess.run(["wc", "-lc", chat_path],
                                    capture_output=True,
                                    text=True)
                n_lines, n_bytes = wc.stdout.split()[:2]
                print(
                    f"  wc -lc          : {n_lines} messages, {n_bytes} bytes")
                print(
                    f"  stat after read : {os.stat(chat_path).st_size} bytes")
                print(f"\n--- open() + read {target}/chat.jsonl ---")
                with open(chat_path) as f:
                    text = f.read().strip()
                if text:
                    for i, line in enumerate(text.splitlines()):
                        if i >= 5:
                            print("  ...")
                            break
                        line = line.strip()
                        if not line:
                            continue
                        try:
                            msg = json.loads(line)
                        except json.JSONDecodeError:
                            break
                        author = msg.get("author", {}).get("username", "?")
                        content = msg.get("content", "")
                        print(f"  [{author}] {content[:80]}")
                else:
                    print("  (empty — no messages on this date)")
                # list attachments
                files_dir = f"{date_dir}/files"
                try:
                    atts = os.listdir(files_dir)
                except (FileNotFoundError, OSError):
                    atts = []
                if atts:
                    print(f"\n--- os.listdir() {target}/files ---")
                    for a in atts[:5]:
                        print(f"  {a}")

        # ── list members ─────────────────────────
        print(f"\n--- os.listdir() {guild}/members ---")
        members = os.listdir(f"{mp}/{guild}/members")
        for m in members:
            print(f"  {m}")

        if members:
            member_path = f"{mp}/{guild}/members/{members[0]}"
            print(f"\n--- open() + read {members[0]} ---")
            with open(member_path) as f:
                text = f.read().strip()
            if text:
                try:
                    data = json.loads(text)
                    user = data.get("user", {})
                    print(f"  username: {user.get('username')}")
                    print(f"  id: {user.get('id')}")
                except json.JSONDecodeError:
                    print(f"  (raw: {text[:100]})")
            else:
                print("  (empty)")

    # ── interactive: browse the mount in another terminal ──
    print(f"\n>>> FUSE mounted at: {mp}")
    print(">>> Open another terminal and run:")
    print(f">>>   ls {mp}/")
    print(f">>>   cat {mp}/<guild>/channels/<ch>/<date>/chat.jsonl")
    print(">>> Press Enter to unmount and exit...")
    input()

    # ── stats ────────────────────────────────────
    records = ws.ops.records
    total = sum(r.bytes for r in records)
    print(f"\nStats: {len(records)} ops, {total} bytes transferred")
